lib/g3-types/src/acl/mod.rs

```diff
@@ -40,7 +40,19 @@ pub use proxy_request::AclProxyRequestRule;
 pub use regex_set::{AclRegexSetRule, AclRegexSetRuleBuilder};
 pub use user_agent::AclUserAgentRule;
 
-#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd)]
+pub trait ActionContract: Clone + PartialEq + Eq + PartialOrd + std::hash::Hash {
+    fn default_forbid() -> Self;
+    fn default_permit() -> Self;
+
+    fn restrict(&self, other: &Self) -> Self;
+    fn strict_than(&self, other: &Self) -> bool;
+    fn forbid_early(&self) -> bool;
```

Comment:
For the most part yes,
except that `restrict` is currently used in

```rust
impl<Action: ActionContract> AclDstHostRuleSetBuilder<Action> {
    pub fn build(&self) -> AclDstHostRuleSet<Action> {
        let mut missed_action = self.missing_action.unwrap_or_else(Action::default_permit);

        let exact_rule = self.exact.as_ref().map(|rule| {
            missed_action = missed_action.restrict(rule.missed_action());
            rule.clone()
        });

        let child_rule = self.child.as_ref().map(|builder| {
            missed_action = missed_action.restrict(builder.missed_action());
            builder.build()
        });

        let regex_rule = self.regex.as_ref().map(|builder| {
            missed_action = missed_action.restrict(builder.missed_action());
            builder.build()
        });

        let subnet_rule = self.subnet.as_ref().map(|builder| {
            missed_action = missed_action.restrict(builder.missed_action());
            builder.build()
        });

        AclDstHostRuleSet {
            exact: exact_rule,
            child: child_rule,
            regex: regex_rule,
            subnet: subnet_rule,
            missed_action,
        }
    }
}
```

Might be possible to do it without that though.

Reply:
I think we can get rid of that by using `.max` that is implemented in `std::cmp` for `Ord`, which it can be. Will do it like that and let me know if you do not like it :)